### PR TITLE
[wip] UDN L2/L3 pod2egress LGW

### PR DESCRIPTION
**Pull request by pperiyasamy**

### CODE REVIEW

**Review by pperiyasamy — COMMENTED** [review]
Thanks a lot @tssurya and @martinkennelly for a very good discussion and great reviews on this PR. Really happy !

File: go-controller/pkg/node/gateway_init.go [hunk 1]
```diff
@@ -477,7 +477,7 @@ func (nc *DefaultNodeNetworkController) initGatewayDPUHost(kubeNodeIP net.IP) er
 		return fmt.Errorf("failed to set the node masquerade route to OVN: %v", err)
 	}
 
-	err = configureSvcRouteViaInterface(nc.routeManager, gatewayIntf, gatewayNextHops)
+	err = configureSvcRouteViaInterface(nc.routeManager, gatewayIntf, gatewayNextHops, 0)
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/gateway_init.go:480]: done.

File: go-controller/pkg/node/gateway_localnet.go [hunk 2]
```diff
@@ -198,6 +199,15 @@ func newLocalGateway(nodeName string, hostSubnets []*net.IPNet, gwNextHops []net
 	return gw, nil
 }
 
+func (g *gateway) programRoutesForUDN(nInfo util.NetInfo, vrfTableId int) error {
+	err := configureSvcRouteViaInterface(g.routeManager, g.GetGatewayBridgeIface(), DummyNextHopIPs(), vrfTableId)
+	if err != nil {
+		return err
+	}
+	// TODO add other routes.
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/gateway_localnet.go:208]: yeah I'd prefer we do it as part of vrf manager, lemme see if i can just compute and pass the routes along.

**Comment by tssurya** [comment]
OK Got things to work functionally... hurray :) there is a LOT more to do ofc like splitting these into smaller PRs and nicer chunks and starting to address some FUP comments like changing vrf manager to not be channel based etc..

**Comment by tssurya** [comment]
adding hold on this PR
I am in the process of splitting this out into more meaningful things

File: go-controller/pkg/node/secondary_node_network_controller.go [hunk 3]
```diff
@@ -54,6 +56,15 @@ func (nc *SecondaryNodeNetworkController) Start(ctx context.Context) error {
 	// Generate a per network conntrack mark to be used for egress traffic.
 	masqCTMark := ctMarkUDNBase + uint(nc.networkID)
 
+	if config.Gateway.Mode == config.GatewayModeLocal {
+		mgmtPortLinkName := util.GetNetMgmtLinkName(nc.NetInfo.GetNetworkName())
+		//TODO; handle returned config, configure ip address
+		_, err := newUserNetManagementPortConfig(mgmtPortLinkName, []*net.IPNet{})
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/secondary_node_network_controller.go]: haha keeping it here was much simpler but let me see if I can move this.

File: go-controller/pkg/node/secondary_node_network_controller.go [hunk 4]
```diff
@@ -54,6 +56,15 @@ func (nc *SecondaryNodeNetworkController) Start(ctx context.Context) error {
 	// Generate a per network conntrack mark to be used for egress traffic.
 	masqCTMark := ctMarkUDNBase + uint(nc.networkID)
 
+	if config.Gateway.Mode == config.GatewayModeLocal {
+		mgmtPortLinkName := util.GetNetMgmtLinkName(nc.NetInfo.GetNetworkName())
+		//TODO; handle returned config, configure ip address
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/secondary_node_network_controller.go]: I have added the port to node annotation.
we are going to skip health check in first version of the PR

File: go-controller/pkg/node/gateway_localnet.go [hunk 5]
```diff
@@ -148,6 +154,46 @@ func newLocalGateway(nodeName string, hostSubnets []*net.IPNet, gwNextHops []net
...
+	gw.wg.Add(1)
+	go func() {
+		defer gw.wg.Done()
+		gw.vrfManager.Run(gw.stopChan, gw.wg)
+	}()
+
+	gw.wg.Add(1)
+	go func() {
+		defer gw.wg.Done()
+		gw.ruleManager.Run(gw.stopChan, 5*time.Minute)
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/gateway_localnet.go]: we still need it actually he he

File: go-controller/pkg/node/gateway_localnet.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/gateway_localnet.go:210]: this is done as part of vrf manager itself

File: go-controller/pkg/ovn/base_network_controller_secondary.go [hunk 6]
```diff
@@ -369,6 +385,42 @@ func (bsnc *BaseSecondaryNetworkController) addLogicalPortToNetworkForNAD(pod *k
 	return nil
 }
 
+func (bsnc *BaseSecondaryNetworkController) buildPodEgressSnatForLGW(podIPs []*net.IPNet, nodeName string) ([]*nbdb.NAT, error) {
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/ovn/base_network_controller_secondary.go]: yup done.

File: go-controller/pkg/node/usernet_management_port.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		"--", "--if-exists", "del-port", "br-int", mpcfg.ifName,
+		"--", "--may-exist", "add-port", "br-int", mpcfg.ifName,
+		"--", "set", "interface", mpcfg.ifName,
+		"type=internal", "mtu_request="+fmt.Sprintf("%d", config.Default.MTU),
+		"external-ids:iface-id="+types.K8sPrefix+mpcfg.ifName)
+	if err != nil {
+		klog.Errorf("Failed to add port to br-int, stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+		return nil, err
+	}
+	macAddress, err := util.GetOVSPortMACAddress(types.K8sMgmtIntfName)
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/usernet_management_port.go]: done.

File: go-controller/pkg/node/usernet_management_port.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		klog.Errorf("Failed to get management port MAC address: %v", err)
+		return nil, err
+	}
+	// persist the MAC address so that upon node reboot we get back the same mac address.
+	_, stderr, err = util.RunOVSVsctl("set", "interface", types.K8sMgmtIntfName,
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/node/usernet_management_port.go]: done.

**Comment by tssurya** [comment]
This PR is going in via:

1. https://github.com/ovn-org/ovn-kubernetes/pull/4543
2. https://github.com/ovn-org/ovn-kubernetes/pull/4546
3. https://github.com/ovn-org/ovn-kubernetes/pull/4548
4. https://github.com/ovn-org/ovn-kubernetes/pull/4549
5. https://github.com/ovn-org/ovn-kubernetes/pull/4550
6. https://github.com/ovn-org/ovn-kubernetes/pull/4552
7. https://github.com/ovn-org/ovn-kubernetes/pull/4553
8. https://github.com/ovn-org/ovn-kubernetes/pull/4554

**Pull request closed without merging**